# Changes made to the UWM TSS/8.25 sources for the Adafruit Fruit Jam build.
# orig/ = bitsavers bits/DEC/pdp8/ascii (and TS8/TS8II from the UWM DECtapes),
# with line endings converted to LF.  lib/dec/ = DEC TSS/8.24 library sources.
# Whitespace and blank-line differences are ignored (diff -u -B -w).
--- orig/SI.PA
+++ src/SI.PA
@@ -3572,11 +3333,10 @@
 PTFORM,	5200		/FORMAT FOR OUTPUT
 PT0260,	260
 PTCOLN,	272		/ASCII COLON
-
 PTDIGI,	0		/ASCII DIGIT BUILT HERE
 PTCONT,	0		/LOOP CONTROL
-
-	IFNZRO TICSPS-24	<GLITCH>
+	IFNZRO TICSPS-24	<IFNZRO TICSPS-12 <GLITCH>>
+	IFZERO TICSPS-24 <
 /DOUBLE PRECISION WORDS...HIGH ORDER BITS
 PTAB,	7520		/-TICKS PER 10 HRS
 	7756		/-TICKS PER 1  HR
@@ -3592,7 +3351,22 @@
 	5520		/-TICKS PER 1 MIN
 	7470		/-TICKS PER 10 SEC
 	7754		/-TICKS PER 1 SEC
-
+	>
+	IFZERO TICSPS-12 <
+/SAME TABLES FOR 10 TICKS PER SECOND (FRUIT JAM)
+PTAB,	7650		/-TICKS PER 10 HRS
+	7767		/-TICKS PER 1  HR
+	7776		/-TICKS PER 10 MIN
+	7777		/-TICKS PER 1  MIN
+	7777		/-TICKS PER 10 SEC
+	7777		/-TICKS PER 1  SEC
+PTAB1,	0700		/-TICKS PER 10 HRS
+	1540		/-TICKS PER 1 HR
+	4220		/-TICKS PER 10 MIN
+	6650		/-TICKS PER 1 MIN
+	7634		/-TICKS PER 10 SEC
+	7766		/-TICKS PER 1 SEC
+	>
 STARS,	215;212;207;"*;"*;"*;" ;0
 QQEST,	"?
 QEST,	"?;0
--- orig/FIP.PA
+++ src/FIP.PA
@@ -167,10 +150,12 @@
 	0
 	OVERLA
 	0
-
 	IFNZRO BILLNG <
 LOGACT,	-BILLNG	/BILLING SYSTEM ACCOUNT NUMBER
 	>
+	IFZERO BILLNG <
+LOGACT,	1	/FRUIT JAM: NO BILLING SYSTEM - MATCHES NO ACCOUNT
+	>
 	*200
 /FILE PHANTOM START
 /FIP'S FIRST JOB IS TO PICK UP THE IOT WHICH IT IS TO PERFORM
@@ -301,11 +284,10 @@
 	LIN
 	LOUT
 	BCLR		/CLEAR ACCOUNT BILLING INFORMATION
+	WHO		/RETURN ACCOUNT & PASSWORD (FRUIT JAM)
 	0
-
 IODSPA,	IODISP-IOTABL-1
 /ROUTINE TO GET A DIRECTORY ENTRY INTO CORE
-
 /CALLING SEQUENCE:
 /	TAD (INTERNAL FILE NUMBER)
 /	JMS GD0
@@ -1885,9 +1788,9 @@
 	LGI0
 	LGO0
 	BCLR0
+	WHO0
 	PAGE
 /ROUTINE TO GET A DIRECTORY WORD INTO CORE
-
 /CALLING SEQUENCE:
 /	TAD (POINTER TO RETRIEVAL INFORMATION)
 /	JMS GE0
@@ -3232,9 +3049,39 @@
 	TAD P0004	/GET THE NUMBER OF PARAMETERS
 	JMP I .+1	/ TO RETURN TO THE USER
 	INF6
-
 OPNNXT,	DSNEXT	/WORD IN 'DS0'
-
+/DEC TSS/8.24 'WHO' IOT (6616): RETURN THE CALLER'S ACCOUNT NUMBER
+/AND PASSWORD.  ADDED FOR THE ADAFRUIT FRUIT JAM BUILD SO THAT THE DEC
+/LIBRARY PROGRAMS (CAT, COPY, GRIPE, LOGOUT) RUN UNCHANGED.
+WHO0,	TAD RETTBL	/MFD RETRIEVAL IS IN ENTRY 0
+	DCA FIOSTK
+	TAD FIACCT
+	DCA FIOSTK+1	/ACCOUNT # TO LOOK FOR
+	CLA CMA		/ONE-WORD SEARCH
+	JMS I DS01
+	 FIOSTK
+	JMP WHOERR	/NOT FOUND - SHOULD NOT HAPPEN
+	DCA INDEX
+	TAD I INDEX
+	DCA FIOSTK+2	/PASSWORD
+	TAD I INDEX
+	DCA FIOSTK+3
+	DATFLD		/SAME AS 'INF5', BUT RETURN ONLY 3 WORDS
+	TAD I FIOPTR	/JOBLNK WORD FROM THE JOB STATUS BLOCK
+	DCA WHODST	/DESTINATION IN FIELD 0
+	CFLD
+	CIF
+	BLT		/MOVE FIOSTK INTO THE IOT PARAMETER BLOCK
+	 CFLD
+	 FIOSTK
+	 DATFLD
+WHODST,	 0
+	 -10
+	TAD P0003
+	JMP I .+1
+	INF6
+WHOERR,	JMP I .+1
+	CPERR2
 	IFZERO .-5200&4000 <GLITCH>
 	*5200
 SEGBUF,				/SEGMENT-SIZE BUFFER
--- orig/PARA.PA
+++ src/PARA.PA
@@ -13,14 +13,13 @@
 	IFNZRO DC08A	<
 		IFNZRO PHONES	<D689=4>
 		IFZERO PHONES	<D689=0>>
-
 NULINE=	DC08A+PT08+KL8	/NUMBER OF TERMINALS LESS CONSOLE
 QUANT=	-6		/NUMBER OF SYSTEM TICKS PER TIME SLICE
 PLT=	0
 STOMIN=	5		/MINIMUM # FREE STORAGE BLOCKS FOR NEW JOB
 SWDEX=	5		/DISC TRACK OF FIRST SWAP AREA
-TICSPS=	24		/NUMBER OF SYSTEM TICKS PER SECOND
-
+TICSPS=	12		/NUMBER OF SYSTEM TICKS PER SECOND (FRUIT JAM: 10/SEC FROM THE
+			/ 60 HZ DK8-E CLOCK, AS IN DEC TSS/8.24; UWM USED 20/SEC)
 	IFZERO RK8E <
 	IFZERO RF08-40 <
 DSKSIZ=	RS08^10
@@ -32,10 +31,14 @@
 DSKSIZ=	400		/# OF 4K DISC TRACKS
 	>
 CLOCKR=	TICSPS%2+1750%TICSPS	/MILLISECONDS PER SYSTEM TICK
+	IFZERO TICSPS-24 <
 INCLK2=	0645		/1,728,000 TICKS PER DAY
+	>
+	IFZERO TICSPS-12 <
+INCLK2=	0322		/864,000 TICKS PER DAY
+	>
 INCLK1=	600^TICSPS	/ (LOW 12 BITS OF 86,400 SECS/DAY * TICKS/SEC)
 TICMIN=	74^TICSPS	/SYSTEM TICKS PER MINUTE
-
 INIDEX=	10	/INIT DISC EXTENSION
 TS8DEX=	30	/FIELD 0 DISC EXTENSION*2^3
 ERPDEX=	0	/ERP DISC EXTENSION*4
@@ -258,6 +233,7 @@
 EXT=	0211	/EXTEND
 RED=	0212	/REDUCE
 FINF=	6613	/FILE INFORMATION
+WHO=	6616	/RETURN ACCOUNT & PASSWORD (DEC 8.24 COMPATIBILITY)
 LIN=	0214	/LOGIN
 LOUT=	0215	/LOGOUT
 BCLR=	0220	/CLEAR MFD BILLING INFORMATION
--- orig/TS8.PA
+++ src/TS8.PA
@@ -1816,17 +1683,14 @@
 	JMS GETUSP	/YES - BUILD THE PARAMETER BLOCK FOR 'FIP'
 	JMP I .+1	/THEN SHARE SOME 'UFILE' CODE TO CALL 'FIP'
 	UUOFIP
-
 UUO3,	TAD UUOT2A
 	DCA UUOSRH	/SET THE ENTRY LIMIT
 	TAD UUOCAL
 	AND C0777
 	JMS UUOSR	/SEARCH FOR THE WHOLE THING
-	JMP I UUOERA	/NOT FOUND - "ILLEGAL IOT"
+	UUOEXT		/NOT FOUND - IGNORE IT, AS DEC TSS/8.24 DOES (FRUIT JAM)
 	JMP UUO1	/YES - FIND THE DISPATCH ADDRESS
-
 /ROUTINE TO PEFORM A BINARY SEARCH ON THE TABLE OF IOTS.
-
 UUOSR,	0
 	CIA
 	DCA UUOLNK
@@ -2402,6 +2238,7 @@
 	613+L	/FINF
 	614	/SIZE
 	615+P	/LOGOUT
+	616+L	/WHO - DEC 8.24 COMPATIBILITY (FRUIT JAM)
 	617	/ACT - ACCOUNT NUMBER
 /
 	620	/BCLR - CLEAR MFD BILLING FIELDS
@@ -2495,6 +2327,7 @@
 	-2	/FINF
 	USIZE
 	0	/LOGOUT
+	-3	/WHO
 	UACT
 /
 	0	/BCLR
--- orig/UWM.PA
+++ src/FRUITJAM.PA
@@ -1,11 +1,10 @@
-/TSS-8.25 CONFIGURATION FOR UWM
+/TSS-8.25 CONFIGURATION FOR UWM - ADAPTED FOR THE ADAFRUIT FRUIT JAM PDP-8/E EMULATOR
 /CONFIGURATION PARAMETERS FOR THE TSS/8 MONITOR
 /COPYRIGHT 1971 - DIGITAL EQUIPMENT CORPORATION; MAYNARD, MASS.
 /
 /TSS-8 SYSTEM PARAMETERS AND OPTIONS.  FOR EACH PARAMETER,
 /THE COMMENT INIDCATES THE POSSIBLE RESPONSES.
-
-DEBUG=	2	/SYSTEM DEBUGGING CONTROL:
+DEBUG=	0	/SYSTEM DEBUGGING CONTROL:
 		/ =0, PRODUCTION SYSTEM (AUTO-RESTARTS & ERROR FIXUP)
 		/ =1, SYSTEM RELIABILITY (JUST AUTO-RESTARTS)
 		/ =2, SYSTEM DEBUGGING (SYSTEM HALTS ON ERROR)
@@ -12,43 +11,27 @@
 		/ =3, FULL CHECKING (THOROUGH 'RETB' CHECK)
 		/BITS 0-8, AVAILABLE FOR LOCAL SWITCHES:
 		/ 4000 - ENABLE MQ MEMORY DISPLAY (IN TIMER)
-
 CPU=	2	/0 IF CPU IS PDP-8/I; 1 IF PDP-8
 		/ 2 IF PDP-8/E; 3 IF PDP-12; 4 IF PDP-8/A
-
 EAE=	0	/20 IF CPU INCLUDES EAE; 0 IF NOT
-
-CORMEM=	50	/HIGHEST DATA FIELD SELECT; FOR 12K SYS,
+CORMEM=	70	/HIGHEST DATA FIELD SELECT; FOR 12K SYS,
 		/ RESPOND 20; FOR 16K, RESPOND 30; FOR 32K, 70
-
 D680=	0	/2 IF TTY INTERFACING IS 680 OR DC08;
 		/ 0 IF TTY INTERFACING IS PT08'S
-
 PHONES=	0	/1 IF 680 OR DC08 INCLUDES 689 OR DC08F;
 		/ 0 IF IT DOESN'T OR TTY INTERFACING IS PT08'S
-
 RF08=	0	/40 IF SYSTEM DISC IS DF32; 0 IF RF08
-
 RS08=	1	/# OF DISCS IN SYSTEM, WHETHER RF08 OR DF32
-
-RK05=	1	/# OF NONSYSTEM RK05 DRIVES
-
+RK05=	0	/# OF NONSYSTEM RK05 DRIVES
 RK8E=	0	/# OF RK8E DRIVES; ONLY FIRST IS USED FOR SYSTEM
 		/ THE REMAINDER ARE TREATED AS PERIPHERALS
-
 READER=	1	/1 IF SYSTEM INCLUDES HIGH-SPEED READER; 0 IF NOT
-
 PUNCH=	1	/1 IF SYS HAS HSP, 0 IF NOT
-
-TC01=	1	/1 IF SYS HAS DECTAPES; 0 IF NOT
-
-
+TC01=	0	/1 IF SYS HAS DECTAPES; 0 IF NOT
 TD8E=	0	/1 IF TD8E DECTAPES AVAILABLE FOR LOADING & DUMPING
 		/ 0 IF NOT; NOTE THAT TD8E TAPES CANNOT BE USED UNDER
 		/ TSS/8, THEY REQUIRE CONTINUOUS CPU ATTENTION
-
-LPT=	3	/0 IF NO PRINTER ON SYS; 1 IF LP08; 2 IF LS8E; 3 IF NONSTANDARD
-
+LPT=	1	/0 IF NO PRINTER ON SYS; 1 IF LP08; 2 IF LS8E; 3 IF NONSTANDARD
 CDR=	0	/NO SYS CAN HAVE CARD READER
 /
 /MONITOR OPTIONS
@@ -54,43 +37,29 @@
 /MONITOR OPTIONS
 /FILL IN THE FOLLOWING LINES TO INDICATE THE
 /MONITOR CAPACITY FOR THIS SYSTEM
-
 	DECIMAL
-JOBMAX=	10	/NUMBER OF SIMULTANEOUS USERS (.LE. 20)
-
+JOBMAX=	20	/NUMBER OF SIMULTANEOUS USERS (.LE. 20)
 CONTTY=	1	/ (ALWAYS HAVE A CONSOLE)
-
 DC08A=	0	/NUMBER OF TERMINALS VIA DC08A (PDP-8/I ONLY, 19 MAX)
-
 DC08LO=	0	/LOWEST LINE-NUMBER USED ON DC08A
-
 PT08=	0	/NUMBER OF TERMINALS VIA PT08 (PDP-8 & PDP-8/I ONLY)
-
-KL8=	15	/NUMBER OF TERMINALS VIA KL8E (PDP-8/E ONLY, 19 MAX)
-
-CDVIN=	3	/NONSTANDARD INPUT CHARACTER DEVICES (HIGHEST USED)
+KL8=	4	/NUMBER OF TERMINALS VIA KL8E (PDP-8/E ONLY, 19 MAX)
+CDVIN=	0	/NONSTANDARD INPUT CHARACTER DEVICES (HIGHEST USED)
 		/ 1 = R2; 2 = X0; 3 = X1; 4 = X2; 5 = X3
-
-CDVOUT=	4	/NONSTANDARD OUTPUT CHARACTER DEVICES (HIGHEST USED)
+CDVOUT=	0	/NONSTANDARD OUTPUT CHARACTER DEVICES (HIGHEST USED)
 		/ 1 = P2; 2 = L2; 3 = Y0; 4 = Y1; 5 = Y2; 6 = Y3
-
-TRUKBD=	10	/NUMBER OF ACTUAL KEYBOARDS ON SYSTEM (NOT COUNTING
+TRUKBD=	4	/NUMBER OF ACTUAL KEYBOARDS ON SYSTEM (NOT COUNTING
 		/ CONSOLE TTY). THIS MUST BE .LE. NULINE; NULINE
 		/ MAY INCLUDE DUMMY TERMINALS, THIS MAY NOT.
-
 FRCMAX=	TRUKBD	/THE HIGHEST KEYBOARD # A '00XX' ACCT # MAY FORCE
 		/ NOTE THAT A '000X' ACCOUNT MAY FORCE UP TO 'NULINE'
-
 TLKMAX=	TRUKBD	/THE HIGHEST KEYBOARD # WHICH MAY BE 'TALK'ED TO
-
 BRDMAX=	TRUKBD	/THE HIGHEST KEYBOARD WHICH WILL RECEIVE BROADCASTS
 	OCTAL
-
-DTAMIN=	2	/THE FIRST DECTAPE DRIVE WHICH WILL BE ASSIGNED
+DTAMIN=	0	/THE FIRST DECTAPE DRIVE WHICH WILL BE ASSIGNED
 		/ FOR 'ASSIGN D' COMMAND; NORMALLY =0, BUT MAY
 		/ VARY FOR SLIGHTLY RESTRICTED DRIVE USE.
-
-BILLNG=	7	/THE ACCOUNT NUMBER OF THE BILLING SYSTEM
+BILLNG=	0	/THE ACCOUNT NUMBER OF THE BILLING SYSTEM
 		/ 0, IF NO BILLING SYSTEM USED
 /
 /SYSTEM CONSTANTS
--- lib/dec/SYSTAT.PA
+++ lib/SYSTAT.PA
@@ -111,7 +111,7 @@
 
 C0002,	2
 CM0002,	-2
-C0022,	22
+C0022,	35	/UWM TSS/8: JOBTBL=DEVTBE+35 (DEC 8.24: +22) - FRUIT JAM
 
 
 INFPA1,	.+1			/PARAMETERS TO MOVE BUFFER TO BLOCK
--- lib/dec/LOGOUT.PA
+++ lib/LOGOUT.PA
@@ -60,7 +60,7 @@
 	CHRBUF=3400	/START OF OUTPUT BUFFER
 	KEYBUF=3600	/START OF INPUT BUFFER
 	UFD=4000		/START OF UFD AREA
-	CLOCK=50	/2 WORD DAY CLOCK IN TS8
+	CLOCK=33	/UWM TSS/8 DAY CLOCK: 33 = LOW ORDER, 34 = HIGH (DEC 8.24: 50/51, HIGH FIRST) - FRUIT JAM
 
 DELB,	0		/# BLOCKS DELETED
 DELC,	0		/# FILES DELETED
@@ -531,7 +531,7 @@
 	RAL
 	AND [7770
 	DCA TIME1	/GET TIME IN DEVICE FORMAT
-	TAD TIME2
+	TAD TIM0	/UWM: LOW ORDER WAS PEEKED INTO TIM0
 	RTL
 	RTL
 	AND [7		/GET LOW ORDER
@@ -588,7 +588,7 @@
 
 TIMPK,	0	/FIELD 0
 	CLOCK
-	TIME1
+	TIM0
 	-2
 /ENTRY AFTER ^C...SHUT OFF ALL TYPEOUTS
 SCTRLC,	TAD	["^
